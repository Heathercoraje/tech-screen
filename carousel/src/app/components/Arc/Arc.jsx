import React, { Component } from "react";
import styled from "styled-components";
import * as d3 from "d3";

export default class Arc extends Component {
  componentDidMount() {
    this.renderArc();
  }
  componentDidUpdate() {
    this.renderArc();
  }

  tau = Math.PI * 2;

  renderArc = () => {
    d3.select("svg").remove();
    const context = this.setContext();
    this.setBackground(context);
    this.setForeground(context);
  };

  setContext = () => {
    let size = this.props.size;
    return d3
      .select(this.refs.arc)
      .append("svg")
      .attr("height", size)
      .attr("width", size)
      .attr("id", "d3-arc")
      .append("g")
      .attr("transform", `translate(${size / 2}, ${size / 2})`);
  };

  arc = () => {
    let size = this.props.size;
    return d3
      .arc()
      .innerRadius(size / 2 - 10)
      .outerRadius(size / 2)
      .startAngle(0);
  };

  setBackground = context => {
    return context
      .append("path")
      .datum({ endAngle: this.tau })
      .style("fill", "transparent")
      .attr("d", this.arc());
  };

  setForeground = context => {
    const score = this.props.score;
    return context
      .append("path")
      .datum({ endAngle: this.tau * this.props.score / 700 })
      .style("fill", "powderblue")
      .attr("d", this.arc());
  };

  render() {
    return <div ref="arc" className="arc" />;
  }
}
