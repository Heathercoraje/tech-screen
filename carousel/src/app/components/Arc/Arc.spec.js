import Arc from "./Arc";
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

describe("Arc", () => {
  const wrapper = shallow(<Arc />);
  it("renders correctly", () => {
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });

  it("should not have style object without props", () => {
    expect(
      wrapper
        .find("div")
        .first()
        .props()
    ).toEqual({ className: "arc" });
  });
  it("should have div component", () => {
    expect(wrapper.find("div").length).toBe(1);
  });
});
