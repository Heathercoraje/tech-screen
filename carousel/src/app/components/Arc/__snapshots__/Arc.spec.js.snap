// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arc renders correctly 1`] = `
<div
  className="arc"
/>
`;
